### PR TITLE
WIP: PoC for `derive(SmartPointer)`

### PR DESCRIPTION
**Pull request by wedsonaf**
Tracking issue: #123430

### CODE REVIEW

**Comment by compiler-errors** [comment]
Marking this as waiting-on-author since it's in draft mode

@rustbot author

**Comment by wedsonaf** [comment]
Rebased.

@rustbot ready

File: library/core/src/lib.rs [hunk 1]
```diff
@@ -475,3 +475,12 @@ pub mod simd {
...
 include!("primitive_docs.rs");
+
+/// Derive macro generating impls of traits related to smart pointers.
+#[cfg(not(bootstrap))]
+#[rustc_builtin_macro]
+#[allow_internal_unstable(dispatch_from_dyn, coerce_unsized, receiver_trait, unsize)]
+#[unstable(feature = "derive_smart_pointer", issue = "123430")]
+pub macro SmartPointer($item:item) {
+    /* compiler built-in */
+}
```
Comment (compiler-errors) [inline library/core/src/lib.rs:486]: This macro definitely should not go into the root module of the core crate. Please look at other derive macros to see what the precedent is.

Comment (pnkfelix) [reply library/core/src/lib.rs:486]: @rustbot author

**Comment by ojeda** [comment]
I have a couple links in Rust for Linux's [#2](https://github.com/Rust-for-Linux/linux/issues/2):

  - Rust Zulip discussion: https://rust-lang.zulipchat.com/#narrow/stream/410673-t-lang.2Fmeetings/topic/RfL.20meeting.202024-03-06/near/431156003.
  - Rust Zulip discussion: https://rust-lang.zulipchat.com/#narrow/stream/425075-rust-for-linux/topic/meeting.202024-04-03/near/431148847.

**Comment by Dylan-DPC** [comment]
@wedsonaf any updates on this? thanks

**Comment by ojeda** [comment]
Linking the RFC PR too: https://github.com/rust-lang/rfcs/pull/3621.

**Comment by Darksonn** [comment]
This should probably be closed as superseded by #125575.

**Pull request closed without merging**